iothub/device/samples/convention-based-samples/TemperatureController/TemperatureControllerSample.cs

```diff
@@ -0,0 +1,375 @@
+                if (temperatureReset)
+                {
+                    // Generate a random value between 5.0°C and 45.0°C for the current temperature reading for each "Thermostat" component.
+                    _temperature[Thermostat1] = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
+                    _temperature[Thermostat2] = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
+                }
+
+                await SendTemperatureAsync(Thermostat1, cancellationToken);
+                await SendTemperatureAsync(Thermostat2, cancellationToken);
+                await SendDeviceMemoryAsync(cancellationToken);
+
+                temperatureReset = _temperature[Thermostat1] == 0 && _temperature[Thermostat2] == 0;
+                await Task.Delay(5 * 1000);
+            }
+        }
+
+        // The callback to handle property update requests.
+        private async Task HandlePropertyUpdatesAsync(ClientPropertyCollection writableProperties, object userContext)
+        {
+            foreach (KeyValuePair<string, object> writableProperty in writableProperties)
+            {
+                switch (writableProperty.Key)
+                {
+                    case Thermostat1:
+                    case Thermostat2:
+                        const string targetTemperatureProperty = "targetTemperature";
+                        if (writableProperties.TryGetValue(writableProperty.Key, targetTemperatureProperty, out double targetTemperatureRequested))
+                        {
+                            await HandleTargetTemperatureUpdateRequestAsync(writableProperty.Key, targetTemperatureRequested, writableProperties.Version, userContext);
+                            break;
+                        }
+                        else
+                        {
+                            _logger.LogWarning($"Property: Received an unrecognized property update from service for component {writableProperty.Key}:" +
+                                        $"\n{{ {writableProperty.Value} }}.");
+                            break;
+                        }
+
+                    default:
+                        _logger.LogWarning($"Property: Received an unrecognized property update from service:" +
+                                        $"\n{{ {writableProperty.Key}: {writableProperty.Value} }}.");
+                        break;
+                }
+            }
+        }
+
+        private async Task HandleTargetTemperatureUpdateRequestAsync(string componentName, double targetTemperature, long version, object userContext)
+        {
+            const string targetTemperatureProperty = "targetTemperature";
+            _logger.LogDebug($"Property: Received - component=\"{componentName}\", {{ \"{targetTemperatureProperty}\": {targetTemperature}°C }}.");
+
+            // Update Temperature in 2 steps
+            double step = (targetTemperature - _temperature[componentName]) / 2d;
+            for (int i = 1; i <= 2; i++)
```

Comment:
If you are executing twice because `_temperature` array is of size 2, why not base it off the size of that array? Safer, no?

Reply:
The fact that we're incrementing the value in "2" steps is a randomly chosen value; just to simulate an actual temperature change event (instead of it going from 15 (current) to 25 (desired) directly). It actually doesn't have any relation to the size of the dict (no. of components).

Reply:
What is the purpose of updating it in two steps? To show it could be long-running?

Reply:
yeah, it'll simulate scenarios where you perform some action within your callback, have delays etc; so a bit closer to how a real device might behave.